integrations/gitlab2/main.py

```diff
@@ -0,0 +1,207 @@
+
+
+# The same sync logic can be registered for one of the kinds that are available in the mapping in port.
+# @ocean.on_resync('project')
+# async def resync_project(kind: str) -> list[dict[Any, Any]]:
+#     # 1. Get all projects from the source system
+#     # 2. Return a list of dictionaries with the raw data of the state
+#     return [{"some_project_key": "someProjectValue", ...}]
+#
+# @ocean.on_resync('issues')
+# async def resync_issues(kind: str) -> list[dict[Any, Any]]:
+#     # 1. Get all issues from the source system
+#     # 2. Return a list of dictionaries with the raw data of the state
+#     return [{"some_issue_key": "someIssueValue", ...}]
+
+
+# Optional
+# Listen to the start event of the integration. Called once when the integration starts.
+@ocean.on_start()
+async def on_start() -> None:
+    # Something to do when the integration starts
+    # For example create a client to query 3rd party services - GitHub, Jira, etc...
+    print("Starting gitlab2 integration")
+
+@ocean.on_resync(ObjectKind.GROUP)
+async def on_group_resync(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE :
+    gitlab_client = init_client()
+
+    async for groups in gitlab_client.get_groups():
+        yield groups
+
+@ocean.on_resync(ObjectKind.PROJECT)
+async def on_project_resync(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE:
+    gitlab_client = init_client()
+    webhook_url = ""
+    semaphore = asyncio.Semaphore(10)
+
+
+    async for projects in gitlab_client.get_projects():
+        logger.info(f"Received project batch with {len(projects)}")
+        tasks = [
+            create_webhook_with_semaphore(
+                semaphore,
+                gitlab_client,
+                project['id'],
+                webhook_url
+            )
+            for project in projects
+        ]
+
+        results = await asyncio.gather(*tasks, return_exceptions=True)
+
+        for result in results:
+            if isinstance(result, Exception):
+                logger.error(f"Error occurred during webhook creation: {result}")
+
+        yield projects
+
+@ocean.on_resync(ObjectKind.MERGE_REQUEST)
+async def on_merge_request_resync(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE :
+    gitlab_client = init_client()
+    async for merge_requests in gitlab_client.get_merge_requests():
+        yield merge_requests
+
+@ocean.on_resync(ObjectKind.ISSUE)
+async def on_issue_resync(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE :
+    gitlab_client = init_client()
+    async for issues in gitlab_client.get_issues():
+        logger.info(f"Received issues batch with {len(issues)}")
+        yield issues
+
+
+@ocean.router.post("/webhook")
+async def handle_webhook(request: Request):
+    event_type = request.headers.get('X-Gitlab-Event')
+    payload = await request.json()
+    await process_webhook_event(event_type, payload)
+
+
+async def create_webhook_with_semaphore(semaphore: asyncio.Semaphore, gitlab_client: GitlabClient, project_id: int, webhook_url: str ):
```

Comment:
Utility functions should either be part of your client class or in the client file